src/utils/ExportData.js

```diff
@@ -0,0 +1,193 @@
+              keyElement2.setAttribute('id', `${key}X`);
+              keyElement2.setAttribute('attr.name', `${key}X`);
+              keyElement2.setAttribute('attr.type', 'double');
+              keyElement2.setAttribute('for', type);
+              graphML.insertBefore(keyElement2, graph);
+            } else {
+              keyElement.setAttribute('attr.type', 'string');
+            }
+            break;
+          default:
+            keyElement.setAttribute('attr.type', 'string');
+        }
+        keyElement.setAttribute('for', type);
+        graphML.insertBefore(keyElement, graph);
+        done.push(key);
+      }
+    });
+  });
+};
+
+const getDataElement = (uri, key, text) => {
+  const data = document.createElementNS(uri, 'data');
+  data.setAttribute('key', key);
+  data.appendChild(document.createTextNode(text));
+  return data;
+};
+
+const addElements = (graph, uri, dataList, type, excludeList, extra: false) => {
+  dataList.forEach((dataElement, index) => {
+    const domElement = document.createElementNS(uri, type);
+    if (dataElement.id) {
+      domElement.setAttribute('id', dataElement.id);
+    } else {
+      domElement.setAttribute('id', index);
+    }
+    if (extra) domElement.setAttribute('source', dataElement.from);
+    if (extra) domElement.setAttribute('target', dataElement.to);
+    graph.appendChild(domElement);
+
+    Object.keys(dataElement).forEach((key) => {
+      if (!excludeList.includes(key)) {
+        if (typeof dataElement[key] !== 'object') {
+          domElement.appendChild(getDataElement(uri, key, dataElement[key]));
+        } else if (dataElement[key].type && dataElement[key].type === 'layout') {
+          domElement.appendChild(getDataElement(uri, `${key}X`, dataElement[key].x));
+          domElement.appendChild(getDataElement(uri, `${key}Y`, dataElement[key].y));
+        } else {
+          domElement.appendChild(getDataElement(uri, key, JSON.stringify(dataElement[key])));
+        }
+      }
+    });
+  });
+};
+
+const xmlToString = (xmlData) => {
+  let xmlString;
+  if (window.ActiveXObject) { // IE
+    xmlString = xmlData.xml;
+  } else { // code for Mozilla, Firefox, Opera, etc.
+    xmlString = (new XMLSerializer()).serializeToString(xmlData);
+  }
+  return xmlString;
+};
+
+const saveFile = (data) => {
+  if (isElectron()) { // electron save dialog
+    const fs = window.require('fs');
+    const { dialog } = window.require('electron').remote;
+    dialog.showSaveDialog({ filters: [{ name: 'graphml', extensions: ['graphml'] }] }, (filename) => {
+      if (filename === undefined) return;
+      fs.writeFile(filename, data, (err) => {
+        if (err) {
+          dialog.showErrorBox('Error Saving File.', err.message);
+        }
+      });
+    });
+  } else if (isCordova()) { // cordova save to downloads
+    window.requestFileSystem(window.LocalFileSystem.TEMPORARY, 5 * 1024 * 1024, (fileSystem) => {
```

Comment:
Elsewhere I have used Promises without bluebird. Is there a reason to use it? Isn't it supported in our target platforms?